ArduPlane/pullup.cpp

```diff
@@ -0,0 +1,244 @@
+    // @Values: 0:Disabled, 1:Enabled
+    // @User: Advanced
+    AP_GROUPINFO_FLAGS("ENABLE", 1, GliderPullup,  enable, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: ELEV_OFS
+    // @DisplayName: Elevator deflection used before starting pullup
+    // @Description: Elevator deflection offset from -1 to 1 while waiting for airspeed to rise before starting close loop control of the pullup.
+    // @Range: -1.0 1.0
+    // @User: Advanced
+    AP_GROUPINFO("ELEV_OFS", 2, GliderPullup,  elev_offset, 0.1f),
+
+    // @Param: NG_LIM
+    // @DisplayName: Maximum normal load factor during pullup
+    // @Description: This is the nominal maximum value of normal load factor used during the closed loop pitch rate control of the pullup.
+    // @Range: 1.0 4.0
+    // @User: Advanced
+    AP_GROUPINFO("NG_LIM", 3, GliderPullup,  ng_limit, 2.0f),
+
+    // @Param: NG_JERK_LIM
+    // @DisplayName: Maximum normal load factor rate of change during pullup
+    // @Description: The normal load factor used for closed loop pitch rate control of the pullup will be ramped up to the value set by PUP_NG_LIM at the rate of change set by this parameter. The parameter value specified will be scaled internally by 1/EAS2TAS.
+    // @Units: 1/s
+    // @Range: 0.1 10.0
+    // @User: Advanced
+    AP_GROUPINFO("NG_JERK_LIM", 4, GliderPullup,  ng_jerk_limit, 4.0f),
+
+    // @Param: PITCH
+    // @DisplayName: Target pitch angle during pullup
+    // @Description: The vehicle will attempt achieve this pitch angle during the pull-up maneouvre.
+    // @Units: deg
+    // @Range: -5 15
+    // @User: Advanced
+    AP_GROUPINFO("PITCH", 5, GliderPullup,  pitch_dem, 3),
+
+    // @Param: ARSPD_START
+    // @DisplayName: Pullup target airspeed
+    // @Description: Target airspeed for initial airspeed wait
+    // @Units: m/s
+    // @Range: 0 100
+    // @User: Advanced
+    AP_GROUPINFO("ARSPD_START", 6, GliderPullup,  airspeed_start, 30),
+
+    // @Param: PITCH_START
+    // @DisplayName: Pullup target pitch
+    // @Description: Target pitch for initial pullup
+    // @Units: deg
+    // @Range: -90 0
+    // @User: Advanced
+    AP_GROUPINFO("PITCH_START", 7, GliderPullup,  pitch_start, -60),
+    
+    AP_GROUPEND
+};
+
+// constructor
+GliderPullup::GliderPullup(void)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+/*
+  return true if in a pullup manoeuvre at the end of NAV_ALTITUDE_WAIT
+*/
+bool GliderPullup::in_pullup(void) const
+{
+    return plane.control_mode == &plane.mode_auto &&
+        plane.mission.get_current_nav_cmd().id == MAV_CMD_NAV_ALTITUDE_WAIT &&
+        stage != Stage::NONE;
+}
+
+/*
+  start a pullup maneouvre, called when NAV_ALTITUDE_WAIT has reached
+  altitude or exceeded descent rate
+*/
+bool GliderPullup::pullup_start(void)
+{
+    if (enable != 1) {
+        return false;
+    }
+
+    // release balloon
+    SRV_Channels::set_output_scaled(SRV_Channel::k_lift_release, 100);
```

Comment:
I would have expected this to be end of altitude wait rather than the start of pullup.

Reply:
Also I have been tending to add new things with a range of 1.

Reply:
a user could chain together multiple altitude waits, with for example a servo command in between to setup some instrument they need at a particular height. They only definately want to release the lift if they do a pullup
on the range, I don't see any set_range(1) in master? it is still a uint16_t, not a float btw.

Reply:
> I don't see any set_range(1) in master? it is still a uint16_t, not a float btw.

There are a few examples (admittedly all added by me):

https://github.com/ArduPilot/ardupilot/blob/d74050c5e328909dba2f3fe3a52825c8fbce7c1d/libraries/AP_ICEngine/AP_ICEngine.cpp#L206-L208

https://github.com/ArduPilot/ardupilot/blob/d74050c5e328909dba2f3fe3a52825c8fbce7c1d/libraries/AP_Motors/AP_MotorsHeli_Single.cpp#L217

The range value is not a float but the scaled value is. So you can have any integer endpoints with no loss of resolution. So we might as well standardize on 1.